src/pubsub/ua_pubsub_reader.c

```diff
@@ -1242,50 +1220,63 @@ UA_ReaderGroup_decodeAndProcessRT(UA_Server *server, UA_ReaderGroup *readerGroup
     useMembufAlloc();
 #endif
 
-    /* Decrypt */
-#ifdef UA_ENABLE_PUBSUB_ENCRYPTION
     UA_NetworkMessage currentNetworkMessage;
     memset(&currentNetworkMessage, 0, sizeof(UA_NetworkMessage));
-
-    size_t payLoadPosition = 0;
+    size_t offset = 0;
+    /* Decode headers necessary for checking identifier */
     UA_StatusCode rv =
-        UA_NetworkMessage_decodeHeaders(buf, &payLoadPosition, &currentNetworkMessage);
+        UA_NetworkMessage_decodeHeadersRT(buf, &offset, &currentNetworkMessage);
     if(rv != UA_STATUSCODE_GOOD) {
         UA_LOG_WARNING_READERGROUP(&server->config.logger, readerGroup,
                               "PubSub receive. decoding headers failed");
         goto done;
     }
 
-    rv = verifyAndDecryptNetworkMessage(&server->config.logger, buf, &payLoadPosition,
-                                        &currentNetworkMessage, readerGroup);
-    UA_NetworkMessage_clear(&currentNetworkMessage);
-    if(rv != UA_STATUSCODE_GOOD) {
-        UA_LOG_WARNING_READERGROUP(&server->config.logger, readerGroup,
-                              "Subscribe failed. verify and decrypt network "
-                              "message failed.");
-        goto done;
-    }
-#endif
-
     /* Process for all readers in the ReaderGroup. Each reader might wait for a
      * different PublisherId. So the offsets are different for each. */
     UA_DataSetReader *dsr;
     LIST_FOREACH(dsr, &readerGroup->readers, listEntry) {
-        decodeAndProcessRT(server, readerGroup, dsr, buf);
+        /* Check the message is intended for this reader */
+        rv = UA_DataSetReader_checkIdentifier(server, &currentNetworkMessage, dsr, readerGroup->config);
+        if(rv == UA_STATUSCODE_GOOD) {
+#ifdef UA_ENABLE_PUBSUB_ENCRYPTION
+            UA_NetworkMessage *nm = dsr->bufferedMessage.nm;
+            size_t payLoadPosition = 0;
+            if(!nm) {
+               /* first call */
+               nm = &currentNetworkMessage;
+            }
+            rv = UA_NetworkMessage_decodeHeaders(buf, &payLoadPosition, nm);
+            if(rv != UA_STATUSCODE_GOOD) {
+                UA_LOG_WARNING_READERGROUP(&server->config.logger, readerGroup,
+                                      "PubSub receive. decoding headers failed");
+                goto done;
+            }
+            /* Decrypt */
+            rv = verifyAndDecryptNetworkMessage(&server->config.logger, buf, &payLoadPosition,
```

Comment:
It seems inefficient to decrypt separately for every reader.

Reply:
The header was already decoded for every packet in the original 1.4 implementation. Now the decoding is only done if the reader group matches. Can there be several reader groups subscribing to the same message?
If yes, then the decryption (and decoding) should only be done for the first matching reader. After this the message is decoded and can be used by all readers.

Reply:
There can be several ReaderGroups sitting on top of the same PubSub-Connection.
Then it depends on their configuration which messages they are processing.
So yes, it would be best to decrypt only when there is at least one match.

You could do a first loop to check for a match, then decode and match again to process.

Reply:
I would prefer a boolean flag to avoid double loops.